// ngx-intl/esm5/lib/pipes/intl-number.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Pipe } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil, skip } from 'rxjs/operators';
import { IntlService } from '../intl.service';
/**
 * @author Abdel El-medny
 *
 * Shows numbers based on the Intl API
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
 */
var IntlNumberPipe = /** @class */ (function () {
    function IntlNumberPipe(changeDetectorRef, $intl) {
        var _this = this;
        this.changeDetectorRef = changeDetectorRef;
        this.$intl = $intl;
        this.lastOutput = null;
        this.lastInput = [null, null];
        this.stop$ = new Subject();
        // If the locale changes, we will take the last inputs we were given, will re-calc the output and trigger CD
        //
        this.$intl.getLocale$().pipe(takeUntil(this.stop$), skip(1))
            .subscribe((/**
         * @param {?} locale
         * @return {?}
         */
        function (locale) {
            var _a = tslib_1.__read(_this.lastInput, 2), lastDate = _a[0], lastOpts = _a[1];
            _this.lastOutput = new Intl.NumberFormat(locale, lastOpts).format(lastDate);
            _this.changeDetectorRef.markForCheck();
        }));
    }
    /**
     * @param {?} value
     * @param {?=} opts
     * @return {?}
     */
    IntlNumberPipe.prototype.transform = /**
     * @param {?} value
     * @param {?=} opts
     * @return {?}
     */
    function (value, opts) {
        // If the input hasn't changed, we will send back the last calculated value
        //
        if (this.lastInput.toString() === value.toString()) {
            return this.lastOutput;
        }
        // Storing this for later usage incase locale changes
        //
        this.lastInput = [(/** @type {?} */ (arguments[0])), (/** @type {?} */ (arguments[1]))];
        /** @type {?} */
        var currentLocale = this.$intl.getLocale$().value;
        this.lastOutput = new Intl.NumberFormat(currentLocale, opts).format(value);
        return this.lastOutput;
    };
    /**
     * @return {?}
     */
    IntlNumberPipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.stop$.next();
    };
    IntlNumberPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'intlNumber',
                    pure: false,
                },] }
    ];
    /** @nocollapse */
    IntlNumberPipe.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: IntlService }
    ]; };
    return IntlNumberPipe;
}());
export { IntlNumberPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.lastOutput;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.lastInput;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.stop$;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.$intl;
}
