// ngx-intl/esm2015/lib/intl.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { NGX_INTL_DEFAULT_LOCALE } from './default-locale.token';
import * as i0 from "@angular/core";
import * as i1 from "./default-locale.token";
/**
 * @author Abdel El-medny
 *
 * This service will be the backbone of the intl number and date pipe
 */
export class IntlService {
    /**
     * @param {?} defaultLocale
     */
    constructor(defaultLocale) {
        this.defaultLocale = defaultLocale;
        this.locale$ = new BehaviorSubject(this.defaultLocale);
    }
    /**
     * @param {?} locale
     * @return {?}
     */
    updateLocale(locale) {
        this.locale$.next(locale);
    }
    /**
     * @return {?}
     */
    getLocale$() {
        return this.locale$;
    }
}
IntlService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
IntlService.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [NGX_INTL_DEFAULT_LOCALE,] }] }
];
/** @nocollapse */ IntlService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function IntlService_Factory() { return new IntlService(i0.ɵɵinject(i1.NGX_INTL_DEFAULT_LOCALE)); }, token: IntlService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    IntlService.prototype.locale$;
    /**
     * @type {?}
     * @private
     */
    IntlService.prototype.defaultLocale;
}
