// ngx-intl/esm2015/lib/pipes/intl-number.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Pipe } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil, skip } from 'rxjs/operators';
import { IntlService } from '../intl.service';
/**
 * @author Abdel El-medny
 *
 * Shows numbers based on the Intl API
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
 */
export class IntlNumberPipe {
    /**
     * @param {?} changeDetectorRef
     * @param {?} $intl
     */
    constructor(changeDetectorRef, $intl) {
        this.changeDetectorRef = changeDetectorRef;
        this.$intl = $intl;
        this.lastOutput = null;
        this.lastInput = [null, null];
        this.stop$ = new Subject();
        // If the locale changes, we will take the last inputs we were given, will re-calc the output and trigger CD
        //
        this.$intl.getLocale$().pipe(takeUntil(this.stop$), skip(1))
            .subscribe((/**
         * @param {?} locale
         * @return {?}
         */
        (locale) => {
            const [lastDate, lastOpts] = this.lastInput;
            this.lastOutput = new Intl.NumberFormat(locale, lastOpts).format(lastDate);
            this.changeDetectorRef.markForCheck();
        }));
    }
    /**
     * @param {?} value
     * @param {?=} opts
     * @return {?}
     */
    transform(value, opts) {
        // If the input hasn't changed, we will send back the last calculated value
        //
        if (this.lastInput.toString() === value.toString()) {
            return this.lastOutput;
        }
        // Storing this for later usage incase locale changes
        //
        this.lastInput = [(/** @type {?} */ (arguments[0])), (/** @type {?} */ (arguments[1]))];
        /** @type {?} */
        const currentLocale = this.$intl.getLocale$().value;
        this.lastOutput = new Intl.NumberFormat(currentLocale, opts).format(value);
        return this.lastOutput;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.stop$.next();
    }
}
IntlNumberPipe.decorators = [
    { type: Pipe, args: [{
                name: 'intlNumber',
                pure: false,
            },] }
];
/** @nocollapse */
IntlNumberPipe.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: IntlService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.lastOutput;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.lastInput;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.stop$;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    IntlNumberPipe.prototype.$intl;
}
