// ngx-intl/esm2015/lib/pipes/intl-date.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Pipe } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil, skip } from 'rxjs/operators';
import { IntlService } from '../intl.service';
/**
 * @author Abdel El-medny
 *
 * Shows dates based on the Intl API
 *
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat
 */
export class IntlDatePipe {
    /**
     * @param {?} changeDetectorRef
     * @param {?} $internationalisation
     */
    constructor(changeDetectorRef, $internationalisation) {
        this.changeDetectorRef = changeDetectorRef;
        this.$internationalisation = $internationalisation;
        this.lastOutput = null;
        this.lastInput = [null, null];
        this.stop$ = new Subject();
        // If the locale changes, we will take the last inputs we were given, will re-calc the output and trigger CD
        //
        this.$internationalisation.getLocale$().pipe(takeUntil(this.stop$), skip(1))
            .subscribe((/**
         * @param {?} locale
         * @return {?}
         */
        (locale) => {
            const [lastValue, lastOpts] = this.lastInput;
            /** @type {?} */
            const dateObject = (lastValue && typeof lastValue === 'string') ? new Date((/** @type {?} */ (lastValue))) : (/** @type {?} */ (lastValue));
            this.lastOutput = new Intl.DateTimeFormat(locale, lastOpts).format(dateObject);
            this.changeDetectorRef.markForCheck();
        }));
    }
    /**
     * @param {?} value
     * @param {?=} opts
     * @return {?}
     */
    transform(value, opts) {
        // If the input hasn't changed, we will send back the last calculated value
        //
        if (this.lastInput.toString() === value.toString()) {
            return this.lastOutput;
        }
        // Storing this for later usage incase locale changes
        //
        this.lastInput = [(/** @type {?} */ (arguments[0])), (/** @type {?} */ (arguments[1]))];
        /** @type {?} */
        const currentLocale = this.$internationalisation.getLocale$().value;
        /** @type {?} */
        const dateObject = typeof value === 'string' ? new Date((/** @type {?} */ (value))) : (/** @type {?} */ (value));
        this.lastOutput = new Intl.DateTimeFormat(currentLocale, opts).format(dateObject);
        return this.lastOutput;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.stop$.next();
    }
}
IntlDatePipe.decorators = [
    { type: Pipe, args: [{
                name: 'intlDate',
                pure: false,
            },] }
];
/** @nocollapse */
IntlDatePipe.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: IntlService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    IntlDatePipe.prototype.lastOutput;
    /**
     * @type {?}
     * @private
     */
    IntlDatePipe.prototype.lastInput;
    /**
     * @type {?}
     * @private
     */
    IntlDatePipe.prototype.stop$;
    /**
     * @type {?}
     * @private
     */
    IntlDatePipe.prototype.changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    IntlDatePipe.prototype.$internationalisation;
}
