// ngx-intl/esm5/lib/intl.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { NGX_INTL_DEFAULT_LOCALE } from './default-locale.token';
import * as i0 from "@angular/core";
import * as i1 from "./default-locale.token";
/**
 * @author Abdel El-medny
 *
 * This service will be the backbone of the intl number and date pipe
 */
var IntlService = /** @class */ (function () {
    function IntlService(defaultLocale) {
        this.defaultLocale = defaultLocale;
        this.locale$ = new BehaviorSubject(this.defaultLocale);
    }
    /**
     * @param {?} locale
     * @return {?}
     */
    IntlService.prototype.updateLocale = /**
     * @param {?} locale
     * @return {?}
     */
    function (locale) {
        this.locale$.next(locale);
    };
    /**
     * @return {?}
     */
    IntlService.prototype.getLocale$ = /**
     * @return {?}
     */
    function () {
        return this.locale$;
    };
    IntlService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    IntlService.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [NGX_INTL_DEFAULT_LOCALE,] }] }
    ]; };
    /** @nocollapse */ IntlService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function IntlService_Factory() { return new IntlService(i0.ɵɵinject(i1.NGX_INTL_DEFAULT_LOCALE)); }, token: IntlService, providedIn: "root" });
    return IntlService;
}());
export { IntlService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IntlService.prototype.locale$;
    /**
     * @type {?}
     * @private
     */
    IntlService.prototype.defaultLocale;
}
