// ngx-intl/esm5/lib/ngx-intl.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NGX_INTL_DEFAULT_LOCALE } from './default-locale.token';
import { IntlDatePipe } from './pipes/intl-date.pipe';
import { IntlNumberPipe } from './pipes/intl-number.pipe';
var NgxIntlModule = /** @class */ (function () {
    function NgxIntlModule() {
    }
    /**
     * @param {?} options
     * @return {?}
     */
    NgxIntlModule.forRoot = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: NgxIntlModule,
            providers: [
                (/** @type {?} */ ({ provide: NGX_INTL_DEFAULT_LOCALE, useValue: options.defaultLocale })),
            ]
        };
    };
    NgxIntlModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [IntlDatePipe, IntlNumberPipe],
                    imports: [],
                    exports: [IntlDatePipe, IntlNumberPipe]
                },] }
    ];
    return NgxIntlModule;
}());
export { NgxIntlModule };
